Support web3@^1.0.0-beta.38

As of 1.0.0-beta.38, web3 ships directly with typescript typings (they seem to no longer be published through @types/web3 as of now). Ideally these typings would be used.

After this version of web3, typechain now throws a bunch of errors with the web3-1.0 target:

```
build/wrappers/A.d.ts:4:58 - error TS2307: Cannot find module 'web3/eth/contract'.

4 import Contract, { CustomOptions, contractOptions } from "web3/eth/contract";
                                                           ~~~~~~~~~~~~~~~~~~~

build/wrappers/A.d.ts:5:46 - error TS2307: Cannot find module 'web3/eth/types'.

5 import { TransactionObject, BlockType } from "web3/eth/types";
                                               ~~~~~~~~~~~~~~~~

build/wrappers/A.d.ts:6:10 - error TS2305: Module '"web3/types"' has no exported member 'Callback'.

6 import { Callback, EventLog } from "web3/types";
           ~~~~~~~~

build/wrappers/A.d.ts:6:20 - error TS2305: Module '"/web3/types"' has no exported member 'EventLog'.

6 import { Callback, EventLog } from "web3/types";
                     ~~~~~~~~

build/wrappers/A.d.ts:8:26 - error TS2307: Cannot find module 'web3/providers'.

8 import { Provider } from "web3/providers";
```

--- a/lib/targets/web3/generation.ts
+++ b/lib/targets/web3/generation.ts
@@ -21,52 +21,31 @@ import {
 
 export function codegen(contract: Contract) {
   const template = `
-  import Contract, { CustomOptions, contractOptions } from "web3/eth/contract";
-  import { TransactionObject, BlockType } from "web3/eth/types";
-  import { Callback, EventLog } from "web3/types";
+  import { Contract, ContractOptions, Options } from "web3-eth-contract";
+  import { Block } from "web3-eth";
+  import { provider } from "web3-providers";
+  import { EventLog } from "web3-core";
   import { EventEmitter } from "events";
-  import { Provider } from "web3/providers";
+  import { Callback, TransactionObject } from "./types";
 
-  export class ${contract.name} {
-    constructor(
-        jsonInterface: any[],
-        address?: string,
-        options?: CustomOptions
-    );
-    _address: string;
-    options: contractOptions;
+  export class ${contract.name} extends Contract {
+    constructor(jsonInterface: any[], address?: string, options?: ContractOptions);
     methods: {
       ${contract.constantFunctions.map(generateFunction).join("\n")}
       ${contract.functions.map(generateFunction).join("\n")}
       ${contract.constants.map(generateConstants).join("\n")}
     };
-    deploy(options: {
-        data: string;
-        arguments: any[];
-    }): TransactionObject<Contract>;
     events: {
       ${contract.events.map(generateEvents).join("\n")}
       allEvents: (
           options?: {
               filter?: object;
-              fromBlock?: BlockType;
+              fromBlock?: Block;
               topics?: (null|string)[];
           },
           cb?: Callback<EventLog>
       ) => EventEmitter;
     };
-    getPastEvents(
-        event: string,
-        options?: {
-            filter?: object;
-            fromBlock?: BlockType;
-            toBlock?: BlockType;
-            topics?: (null|string)[];
-        },
-        cb?: Callback<EventLog[]>
-    ): Promise<EventLog[]>;
-    setProvider(provider: Provider): void;
-    clone(): ${contract.name};
 }
   `;
 
@@ -112,7 +91,7 @@ function generateEvents(event: EventDeclaration) {
   ${event.name}(
     options?: {
         filter?: object;
-        fromBlock?: BlockType;
+        fromBlock?: Block;
         topics?: (null|string)[];
     },
     cb?: Callback<EventLog>): EventEmitter;

--- a/lib/targets/web3/generation.ts
+++ b/lib/targets/web3/generation.ts
@@ -23,7 +23,6 @@ export function codegen(contract: Contract) {
   const template = `
   import { Contract, ContractOptions, Options } from "web3-eth-contract";
   import { Block } from "web3-eth";
-  import { provider } from "web3-providers";
   import { EventLog } from "web3-core";
   import { EventEmitter } from "events";
   import { Callback, TransactionObject } from "./types";

--- a/lib/targets/web3/index.ts
+++ b/lib/targets/web3/index.ts
@@ -40,4 +40,24 @@ export class Web3 extends TsGeneratorPlugin {
       contents: codegen(contract),
     };
   }
+
+  afterRun(): TFileDesc[] {
+    return [
+      {
+        path: join(this.outDirAbs, "types.d.ts"),
+        contents: `
+  import { Transaction } from "web3-core";
+  // @ts-ignore
+  import PromiEvent from "web3-core-promievent";
+  export type Callback<T> = (error: Error, result: T) => void;
+  export interface TransactionObject<T> {
+    arguments: any[];
+    call(tx?: Transaction): Promise<T>;
+    send(tx?: Transaction): PromiEvent<T>;
+    estimateGas(tx?: Transaction): Promise<number>;
+    encodeABI(): string;
+  }`,
+      },
+    ];
+  }
 }

--- a/test/integration/targets/web3-1.0.0/DumbContract.spec.web3.ts
+++ b/test/integration/targets/web3-1.0.0/DumbContract.spec.web3.ts
@@ -5,7 +5,7 @@ import { expect } from "chai";
 
 describe("DumbContract", () => {
   it("should work", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract: DumbContract = await deployContract<DumbContract>("DumbContract");
 
     expect(await contract.methods.returnAll().call({ from: accounts[0] })).to.be.deep.eq({
       "0": "0",
@@ -14,13 +14,13 @@ describe("DumbContract", () => {
   });
 
   it("should have an address", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract = await deployContract<DumbContract>("DumbContract");
 
-    expect(await contract._address).to.be.string;
+    expect(await contract.options.address).to.be.string;
   });
 
   it("should allow to pass unsigned values in multiple ways", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract = await deployContract<DumbContract>("DumbContract");
 
     await contract.methods.countup(2).send({ from: accounts[0] });
     expect(await contract.methods.counter().call()).to.be.eq("2");
@@ -29,14 +29,14 @@ describe("DumbContract", () => {
   });
 
   it("should allow to pass signed values in multiple ways", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract = await deployContract<DumbContract>("DumbContract");
 
     expect(await contract.methods.returnSigned(2).call()).to.be.eq("2");
     expect(await contract.methods.returnSigned("2").call()).to.be.eq("2");
   });
 
   it("should allow to pass address values in multiple ways", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract = await deployContract<DumbContract>("DumbContract");
 
     expect(
       await contract.methods.testAddress("0x0000000000000000000000000000000000000123").call(),
@@ -44,34 +44,38 @@ describe("DumbContract", () => {
   });
 
   it("should allow to pass bytes values in multiple ways", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract = await deployContract<DumbContract>("DumbContract");
+    const byteString = "0xabcd123456000000000000000000000000000000000000000000000000000000";
 
-    const res = await contract.methods.callWithBytes([1, 0, 1]).call();
-    expect(res).to.be.deep.eq("0x0100010000000000000000000000000000000000000000000000000000000000");
+    await contract.methods.callWithBytes(byteString).send({ from: accounts[0] });
+
+    const result = await contract.methods.byteArray().call();
+    expect(result).to.be.a("string");
+    expect(result).to.eq(byteString);
   });
 
   it("should allow to pass boolean values", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract = await deployContract<DumbContract>("DumbContract");
 
     const res = await contract.methods.callWithBoolean(true).call();
     expect(res).to.be.deep.eq(true);
   });
 
   it("should allow to pass numeric arrays values in multiple ways", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract = await deployContract<DumbContract>("DumbContract");
 
     const res = await contract.methods.callWithArray2(["1", 2]).call();
     expect(res).to.be.deep.eq(["1", "2"]);
   });
 
   it("should allow to pass strings ", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract = await deployContract<DumbContract>("DumbContract");
 
     expect(await contract.methods.testString("abc").call()).to.be.deep.eq("abc");
   });
 
   it("should allow to clone contracts ", async () => {
-    const contract = (await deployContract("DumbContract")) as DumbContract;
+    const contract = await deployContract<DumbContract>("DumbContract");
 
     const contractClone = await contract.clone();
 

--- a/test/integration/targets/web3-1.0.0/package.json
+++ b/test/integration/targets/web3-1.0.0/package.json
@@ -4,13 +4,12 @@
   "main": "index.js",
   "license": "MIT",
   "dependencies": {
-    "web3": "^1.0.0-beta.36"
+    "web3": "^1.0.0-beta.48"
   },
   "devDependencies": {
     "@types/bn.js": "^4.11.1",
     "@types/chai": "^4.1.2",
     "@types/underscore": "^1.8.9",
-    "@types/web3": "^1.0.4",
     "chai": "^4.1.2",
     "mocha": "^5.2.0",
     "ts-node": "^7.0.1",

--- a/test/integration/targets/web3-1.0.0/tsconfig.json
+++ b/test/integration/targets/web3-1.0.0/tsconfig.json
@@ -6,7 +6,8 @@
     "lib": ["es2015"],
     "target": "es2016",
     "module": "commonjs",
-    "declaration": true
+    "declaration": true,
+    "esModuleInterop": true
   },
   "include": [
     "./**/*.ts"

--- a/test/integration/targets/web3-1.0.0/web3.ts
+++ b/test/integration/targets/web3-1.0.0/web3.ts
@@ -1,6 +1,6 @@
 const ganache = require("ganache-cli");
 
-import Web3 = require("web3");
+import Web3 from "web3";
 import { join } from "path";
 import { readFileSync } from "fs";
 

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -30,6 +30,7 @@
     "./lib/targets/legacy/runtime/**/*.ts",
     "./test/integration/targets/truffle/**/*.ts",
     "./test/integration/targets/web3-1.0.0/**/*",
-    "./test/integration/targets/ethers/**/*"
+    "./test/integration/targets/ethers/**/*",
+    "./test/package-test/**/*"
   ]
 }

--- a/tsconfig.production.json
+++ b/tsconfig.production.json
@@ -17,6 +17,7 @@
   "exclude": [
     "./lib/targets/legacy/runtime/**/*.ts",
     "./test/integration/targets/truffle/**/*.ts",
-    "./test/integration/targets/web3-1.0.0/**/*"
+    "./test/integration/targets/web3-1.0.0/**/*",
+    "./test/package-test/**/*"
   ]
 }

